schema/dashboard.1.sql

```diff
@@ -0,0 +1,32 @@
+CREATE TABLE "meta" (
+    "version" INTEGER NOT NULL
+);
+
+-- dashboard.2.sql should UPDATE "meta" SET "version" = 2; and so on
+INSERT INTO "meta" VALUES (1);
+
+CREATE TABLE "test" (
+    "path" TEXT NOT NULL                        -- test name
+    , "subtest" TEXT
```

Comment:
tests are identified by (path,subtest), but some tests are not subtests. the obvious representation for such tests would be (path,NULL), but sqlite makes the (legal) decision to allow (path,NULL) to effectively bypass

1. PRIMARY KEY constraints ([“Each row in a table with a primary key must have a unique combination of values in its primary key columns. For the purposes of determining the uniqueness of primary key values, NULL values are considered distinct from all other values, including other NULLs.”](https://sqlite.org/lang_createtable.html#the_primary_key))
2. UNIQUE constraints ([“For each UNIQUE constraint on the table, each row must contain a unique combination of values in the columns identified by the UNIQUE constraint. For the purposes of UNIQUE constraints, NULL values are considered distinct from all other values, including other NULLs.”](https://sqlite.org/lang_createtable.html#unique_constraints))
3. FOREIGN KEY constraints ([“The foreign key constraint is satisfied if for each row in the child table either one or more of the child key columns are NULL, or there exists a row in the parent table for which each parent key column contains a value equal to the value in its associated child key column.”](https://sqlite.org/foreignkeys.html#fk_basics))

i managed to solve this by converting (path,NULL) to (path,'') and (path,subtest) to (path,'='||subtest) between the application and the database, but this makes every insert and select a bug waiting to happen. tomorrow i’ll see if i can automate this conversion with a [generated column](https://sqlite.org/gencol.html) that gets used in constraints.

Reply:
i managed to solve this by converting (path,NULL) to (path,'') and (path,subtest) to (path,'='||subtest) between the application and the database (191a6de9203bea8e9eddc27f91f66947ce20f151), but this makes every insert and select a bug waiting to happen.

tomorrow i’ll see if i can automate this conversion with a [generated column](https://sqlite.org/gencol.html) that gets used in constraints.

Reply:
done in 929854f14d2081731d9a99d7cd70065644ed763a